File: internal/anim/spinner.go
// Package anim provides Charm Bubbles–style spinner presets ported from
// the Rust donk-anim crate.
package anim

import "time"

// SpinnerKind identifies a spinner preset.
type SpinnerKind string

const (
	SpinnerKindLine     SpinnerKind = "line"
	SpinnerKindDot      SpinnerKind = "dot"
	SpinnerKindMiniDot  SpinnerKind = "minidot"
	SpinnerKindJump     SpinnerKind = "jump"
	SpinnerKindPulse    SpinnerKind = "pulse"
	SpinnerKindPoints   SpinnerKind = "points"
	SpinnerKindMoon     SpinnerKind = "moon"
)

// SpinnerKinds is the ordered list of available spinner presets.
var SpinnerKinds = []SpinnerKind{
	SpinnerKindLine,
	SpinnerKindDot,
	SpinnerKindMiniDot,
	SpinnerKindJump,
	SpinnerKindPulse,
	SpinnerKindPoints,
	SpinnerKindMoon,
}

// spinnerFrames maps each kind to its frame set.
var spinnerFrames = map[SpinnerKind][]string{
	SpinnerKindLine:    {"|", "/", "-", "\\"},
	SpinnerKindDot:     {"⣾", "⣽", "⣻", "⢿", "⡿", "⣟", "⣯", "⣷"},
	SpinnerKindMiniDot: {"⠋", "⠙", "⠹", "⠸", "⠼", "⠴", "⠦", "⠧", "⠇", "⠏"},
	SpinnerKindJump:    {"⢄", "⢂", "⢁", "⡁", "⡈", "⡐", "⡠"},
	SpinnerKindPulse:   {"▏", "▎", "▍", "▌", "▋", "▊", "▉", "▊", "▋", "▌", "▍", "▎"},
	SpinnerKindPoints:  {"∙∙∙", "●∙∙", "∙●∙", "∙∙●"},
	SpinnerKindMoon:    {"🌑", "🌒", "🌓", "🌔", "🌕", "🌖", "🌗", "🌘"},
}

// spinnerInterval maps each kind to its tick interval.
var spinnerInterval = map[SpinnerKind]time.Duration{
	SpinnerKindLine:    100 * time.Millisecond,
	SpinnerKindDot:     80 * time.Millisecond,
	SpinnerKindMiniDot: 80 * time.Millisecond,
	SpinnerKindJump:    100 * time.Millisecond,
	SpinnerKindPulse:   120 * time.Millisecond,
	SpinnerKindPoints:  160 * time.Millisecond,
	SpinnerKindMoon:    200 * time.Millisecond,
}

// Spinner is a Charm Bubbles–style spinner frame player.
type Spinner struct {
	kind    SpinnerKind
	index   int
	frames  []string
	last    time.Time
	interval time.Duration
}

// NewSpinner creates a spinner with the given preset.
func NewSpinner(kind SpinnerKind) *Spinner {
	frames := spinnerFrames[kind]
	if len(frames) == 0 {
		frames = spinnerFrames[SpinnerKindDot]
		kind = SpinnerKindDot
	}
	return &Spinner{
		kind:     kind,
		index:    0,
		frames:   frames,
		last:     time.Now(),
		interval: spinnerInterval[kind],
	}
}

// Kind returns the active spinner preset.
func (s *Spinner) Kind() SpinnerKind { return s.kind }

// SetKind changes the spinner preset and resets position.
func (s *Spinner) SetKind(kind SpinnerKind) {
	frames := spinnerFrames[kind]
	if len(frames) == 0 {
		return
	}
	s.kind = kind
	s.frames = frames
	s.index = 0
	s.last = time.Now()
	s.interval = spinnerInterval[kind]
}

// CycleKind advances to the next preset.
func (s *Spinner) CycleKind() {
	all := SpinnerKinds
	i := 0
	for j, k := range all {
		if k == s.kind {
			i = j
			break
		}
	}
	s.SetKind(all[(i+1)%len(all)])
}

// Update advances the spinner if its interval has elapsed.
// It is safe to call on every animation tick.
func (s *Spinner) Update(now time.Time) {
	if now.IsZero() {
		now = time.Now()
	}
	if now.Sub(s.last) >= s.interval {
		s.index = (s.index + 1) % len(s.frames)
		s.last = now
	}
}

// Frame returns the current spinner frame string.
func (s *Spinner) Frame() string {
	if len(s.frames) == 0 {
		return ""
	}
	return s.frames[s.index%len(s.frames)]
}

File: internal/ui/anim/library/rust-reference/spinner.rs
//! Charm Bubbles–style spinner frames (port of presets, not a Go runtime).

use ratatui::prelude::*;
use std::time::{Duration, Instant};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SpinnerKind {
    Line,
    Dot,
    MiniDot,
    Jump,
    Pulse,
    Points,
    Moon,
}

impl SpinnerKind {
    pub fn all() -> &'static [SpinnerKind] {
        &[
            Self::Line,
            Self::Dot,
            Self::MiniDot,
            Self::Jump,
            Self::Pulse,
            Self::Points,
            Self::Moon,
        ]
    }

    pub fn next(self) -> Self {
        let all = Self::all();
        let i = all.iter().position(|k| *k == self).unwrap_or(0);
        all[(i + 1) % all.len()]
    }

    pub fn label(self) -> &'static str {
        match self {
            Self::Line => "line",
            Self::Dot => "dot",
            Self::MiniDot => "minidot",
            Self::Jump => "jump",
            Self::Pulse => "pulse",
            Self::Points => "points",
            Self::Moon => "moon",
        }
    }

    fn frames(self) -> &'static [&'static str] {
        match self {
            Self::Line => &["|", "/", "-", "\\"],
            Self::Dot => &["⣾", "⣽", "⣻", "⢿", "⡿", "⣟", "⣯", "⣷"],
            Self::MiniDot => &["⠋", "⠙", "⠹", "⠸", "⠼", "⠴", "⠦", "⠧", "⠇", "⠏"],
            Self::Jump => &["⢄", "⢂", "⢁", "⡁", "⡈", "⡐", "⡠"],
            Self::Pulse => &["▏", "▎", "▍", "▌", "▋", "▊", "▉", "▊", "▋", "▌", "▍", "▎"],
            Self::Points => &["∙∙∙", "●∙∙", "∙●∙", "∙∙●"],
            Self::Moon => &["🌑", "🌒", "🌓", "🌔", "🌕", "🌖", "🌗", "🌘"],
        }
    }

    fn interval(self) -> Duration {
        match self {
            Self::Line => Duration::from_millis(100),
            Self::Dot | Self::MiniDot => Duration::from_millis(80),
            Self::Jump => Duration::from_millis(100),
            Self::Pulse => Duration::from_millis(120),
            Self::Points => Duration::from_millis(160),
            Self::Moon => Duration::from_millis(200),
        }
    }
}

#[derive(Debug, Clone)]
pub struct Spinner {
    kind: SpinnerKind,
    index: usize,
    last: Instant,
}

impl Default for Spinner {
    fn default() -> Self {
        Self::new()
    }
}

impl Spinner {
    pub fn new() -> Self {
        Self::with_kind(SpinnerKind::Dot)
    }

    pub fn with_kind(kind: SpinnerKind) -> Self {
        Self {
            kind,
            index: 0,
            last: Instant::now(),
        }
    }

    pub fn kind(&self) -> SpinnerKind {
        self.kind
    }

    pub fn set_kind(&mut self, kind: SpinnerKind) {
        self.kind = kind;
        self.index = 0;
        self.last = Instant::now();
    }

    pub fn cycle_kind(&mut self) {
        self.set_kind(self.kind.next());
    }

    pub fn update(&mut self) {
        if self.last.elapsed() >= self.kind.interval() {
            let frames = self.kind.frames();
            self.index = (self.index + 1) % frames.len();
            self.last = Instant::now();
        }
    }

    pub fn frame(&self) -> &'static str {
        let frames = self.kind.frames();
        frames[self.index % frames.len()]
    }

    pub fn view_lines(&self) -> Vec<Line<'static>> {
        let preview: String = self
            .kind
            .frames()
            .iter()
            .map(|f| format!("{f} "))
            .collect();
        vec![
            Line::from(Span::styled(
                format!("spinner · {}", self.kind.label()),
                Style::default().add_modifier(Modifier::BOLD),
            )),
            Line::from(""),
            Line::from(Span::styled(
                format!("  {}  ", self.frame()),
                Style::default()
                    .fg(Color::Rgb(0xF9, 0x67, 0xDC))
                    .add_modifier(Modifier::BOLD),
            )),
            Line::from(""),
            Line::from(Span::styled(
                preview,
                Style::default().fg(Color::DarkGray),
            )),
            Line::from(""),
            Line::from("space cycle preset · tab next"),
        ]
    }
}
